### PR TITLE
[GV-49] Setup File Upload

### PR DESCRIPTION
**Pull request by Connor-Bernard**
### Jira Ticket

[Jira Ticket](https://dashbd.atlassian.net/browse/GV-49)

### Description

- Add support for CM schema file uploads
- Add index route

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would change existing functionality)

### Changes

- Add admin route
- Add post route for uploading CM schema files
- Add get (index) route for fetching existing CM schema files
- Add file upload helper
- Add custom mime type for CM schema files

### Checklist

- [x] My branch name matches the format: `<ticket-id>/<brief-description-of-change>`
- [x] My PR name matches the format: `[<ticket-id>] <brief-description-of-change>`
- [x] I have added doc-comments to all new functions ([JSDoc](https://jsdoc.app/) for JS and [Docstrings](https://peps.python.org/pep-0257/) for Python)
- [x] I have reviewed all of my code
- [x] My code only contains major changes related to my ticket

### CODE REVIEW

File: api/lib/uploadHandler.mjs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                        uploadedAt: new Date(),
+                    };
+                    console.log(
+                        '[LOG]: Accepted file upload:',
+                        req.fileMetadata,
+                    );
+                    return next();
+                }
+                if (req.file) {
+                    unlink(req.file.path, (unlinkError) => {
```
[the remark this answers is not in the document]
Comment (Connor-Bernard) [reply api/lib/uploadHandler.mjs]: smh this is a rly good point.  I should have looked into ulink a bit more.  I thought it was gated to the open file buffer 🙄

[the remark this answers is not in the document]
Comment (Connor-Bernard) [reply api/lib/uploadHandler.mjs]: Resolved in [b813173](https://github.com/AFA-Tooling/GradeView/pull/15/commits/b8131733199ca2a72db80120810f18cbd65ef1dd)

File: api/lib/uploadHandler.mjs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+     * @param {string} mimeType the mime type of the file.
+     * @param {number} fileSize the maximum number of bytes allowed.
+     * @param {function} fileNameTransformer the cb for the file name.
+     */
+    constructor(
+        fieldName,
+        directory,
+        mimeType,
+        fileSize,
+        fileNameTransformer = (f) => sanitize(f.originalname),
```
Comment (Mehul-Gandhi) [inline api/lib/uploadHandler.mjs:27]: What is the plan for guidelines on what the fileName can be? What if school A creates the file `cs1.cm`  and school B also creates the file `cs1.cm`? Then are we overwriting school A's `cs1.cm` file? Do we want to add more metadata for the fileName to ensure files are not overwritten? Are we saving these files locally? What if we restart the application, then does the data persist? Are we going to move this to a database later?

Comment (Mehul-Gandhi) [reply api/lib/uploadHandler.mjs:27]: If we are just supporting CS10 for now, then this duplicate naming convention will probably not be a problem. But we should recommend some naming convention in the README or the user interface, for example `cs10_{semester}{year}.cm`

File: api/v2/Routes/admin/progressReports/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    // TODO: move over to the compiled schema folder when created.
+    res.json(
+        Array.from(schemaFiles).map((fileName) =>
+            fileName.substring(0, fileName.lastIndexOf('.')),
+        ),
+    );
+});
+
+router.get('/:schemaName', (_, res) => {
+    // We need to have the CM schema parsing endpoint set up to support this.
```
Comment (Mehul-Gandhi) [inline api/v2/Routes/admin/progressReports/index.js:34]: Is this a GET request to retrieve a particular file stored?

File: api/lib/uploadHandler.mjs (hunk 2 above)
Comment (Connor-Bernard) [reply api/lib/uploadHandler.mjs:27]: 1. Has to be less than 100 chars
2. Allowing overwrites
3. Maybe in the future
4. Stored as files in directory
5. Data persists
6. Ideally
7. naming conventions would be optimal, but it is what the user needs to input in the route, so I'm leaving it configurable to the admin

Comment (Connor-Bernard) [reply api/lib/uploadHandler.mjs:27]: Once we have a db we can map it to a key and call it a day

File: api/v2/Routes/admin/progressReports/index.js (hunk 3 above)
Comment (Connor-Bernard) [reply api/v2/Routes/admin/progressReports/index.js:34]: Yep

**Review by Mehul-Gandhi — APPROVED** [review]
LGTM

**Pull request merged**